src/utilities.jl

```diff
@@ -52,13 +62,107 @@ end
+##### read/write/bytes/streams
+#####
 
 jump(io::IO, n) = (read(io, n); nothing)
 jump(io::IOStream, n) = (skip(io, n); nothing)
 jump(io::IOBuffer, n) = ((io.seekable ? skip(io, n) : read(io, n)); nothing)
 
 unsafe_vec_uint8(x::AbstractVector{UInt8}) = convert(Vector{UInt8}, x)
 unsafe_vec_uint8(x::Base.ReinterpretArray{UInt8,1}) = unsafe_wrap(Vector{UInt8}, pointer(x), length(x))
+
+"""
+    read_byte_range(path, byte_offset, byte_count)
+
+Return the equivalent `read(path)[(byte_offset + 1):(byte_offset + byte_count)]`,
+but try to avoid reading unreturned intermediate bytes. Note that the
+effectiveness of this method depends on the type of `path`.
+"""
+function read_byte_range(path, byte_offset, byte_count)
+    return open(path, "r") do io
+        jump(io, byte_offset)
+        return read(io, byte_count)
+    end
+end
+
+read_byte_range(path, ::Missing, ::Missing) = read(path)
+
+write_full_path(path::AbstractString, bytes) = (mkpath(dirname(path)); write(path, bytes))
+write_full_path(path, bytes) = write(path, bytes)
+
+#####
+##### tables
+#####
+
+function table_has_supported_onda_format_version(table)
+    m = Arrow.getmetadata(table)
+    return m isa Dict && is_supported_onda_format_version(VersionNumber(get(m, "onda_format_version", v"0.0.0")))
+end
+
+# It would be better if Arrow.jl supported a generic API for nonstandard path-like types so that
+# we can avoid potential intermediate copies here, but its documentation is explicit that it only
+# supports `Union{IO,String}`.
+
+read_arrow_table(io_or_path::Union{IO,String,Vector{UInt8}}) = Arrow.Table(io_or_path)
+read_arrow_table(path) = read_arrow_table(read(path))
+
+write_arrow_table(path::String, table; kwargs...) = Arrow.write(path, table; kwargs...)
+write_arrow_table(io::IO, table; kwargs...) = Arrow.write(io, table; file=true, kwargs...)
+write_arrow_table(path, table; kwargs...) = (io = IOBuffer(); write_arrow_table(io, table; kwargs...); write_full_path(path, take!(io)))
+
+function read_onda_table(path; materialize::Bool=false)
+    table = read_arrow_table(path)
+    table_has_supported_onda_format_version(table) || error("supported `onda_format_version` not found in annotations file")
+    return materialize ? map(collect, Tables.columntable(table)) : table
+end
+
+function write_onda_table(path, table; kwargs...)
+    Arrow.setmetadata!(table, Dict("onda_format_version" => "v$(MAXIMUM_ONDA_FORMAT_VERSION)"))
+    write_arrow_table(path, table; kwargs...)
+    return table
+end
+
+function locations(collections::NTuple{N}) where {N}
+    K = promote_type(eltype.(collections)...)
+    results = Dict{K,NTuple{N,Vector{Int}}}()
+    for (c, collection) in enumerate(collections)
+        for (i, item) in enumerate(collection)
+            push!(get!(() -> ntuple(_ -> Int[], N), results, item)[c], i)
+        end
+    end
+    return results
+end
+
+function _iterator_for_column(table, c)
+    Tables.columnaccess(table) && Tables.getcolumn(Tables.columns(table), c)
```

Comment:
```suggestion
    Tables.columnaccess(table) && return Tables.getcolumn(Tables.columns(table), c)
```

Reply:
Oh jeez, nice catch - I should make sure this code path is tested for sure.